docs/dev/sql-nested-function-select-clause.md

```diff
@@ -0,0 +1,413 @@
+
+        phyState1 --> phyState2
+        phyState2 --> IndexScan
+    }
+```
+
+### 2.2 Sequence Diagram for Nested SELECT Clause Query Push Down
+Nested function sequence diagram illustrating query execution from parsing to OpenSearch DSL push down.
+
+```mermaid
+sequenceDiagram
+    participant SQLService
+    participant ParserBaseRoot
+    participant AstExpressionBuilder
+
+    participant QueryService
+    participant Analyzer
+    participant NestedAnalyzer
+    participant Planner
+    participant TableScanPushDown
+    participant OpenSearchRequestBuilder
+    participant DefaultImplementor
+
+%% Parsing
+SQLService->>ParserBaseRoot:visitRoot
+  ParserBaseRoot->>AstExpressionBuilder:visitScalarFunction
+  AstExpressionBuilder-->>ParserBaseRoot:Function
+ParserBaseRoot-->>SQLService:UnresolvedPlan
+%% Analysis
+SQLService->>QueryService:analyze
+  QueryService->>Analyzer:visitProject
+    Analyzer->>NestedAnalyzer:visitFunction
+      NestedAnalyzer-->>Analyzer:LogicalNested
+    Analyzer-->>QueryService:UnresolvedPlan
+  QueryService->>Planner:plan
+    %% planner optimization
+    Planner->>TableScanPushDown:apply
+      TableScanPushDown->>OpenSearchRequestBuilder:pushDownNested
+      OpenSearchRequestBuilder-->>TableScanPushDown:boolean
+    TableScanPushDown-->>Planner:LogicalPlan
+    %% planner implementation
+    Planner->>DefaultImplementor:visitNested
+    DefaultImplementor-->>Planner:NestedOperator
+  Planner-->>QueryService:PhysicalPlan
+QueryService-->>SQLService:PhysicalPlan
+```
+
+### 2.3 Sequence Diagram for Nested SELECT Clause Post-processing
+Nested function sequence diagram illustrating the flattening of the OpenSearch response. Flattening the response from OpenSearch changes the nested types structure by making the full path of an object the key, and the object it refers to the value.
+
+**Sample input:**
+```json
+{
+  "comments": {
+    "likes": 2
+  }
+}
+```
+**Sample Output:**
+```json
+{
+  "comment.likes": 2
+}
+```
+
+```mermaid
+sequenceDiagram
+
+%% Flattening
+ProjectOperator->>ResourceMonitorPlan:next
+  ResourceMonitorPlan->>NestedOperator:next
+    loop unnesting
+      NestedOperator->>NestedOperator:flatten
+    end
+  NestedOperator-->>ResourceMonitorPlan:ExprValue
+ResourceMonitorPlan-->>ProjectOperator:ExprValue
+
+%% Resolving
+ProjectOperator->>NamedExpression.delegate:valueOf
+  NamedExpression.delegate->>OpenSearchFunctions.nested:resolve
+  OpenSearchFunctions.nested-->>NamedExpression:ExprValue
+NamedExpression-->>ProjectOperator:ExprValue
+```
+
+#### 2.4 Select Clause Nested Query Class Diagram
+Nested function class diagram for additional classes required for query execution. The `NestedAnalyzer` is a visitor for nested functions used in the SELECT clause to fulfill the `LogicalNested` LogicalPlan. After push down is successful the `NestedOperator` PhysicalPlan is used for object flattening of the OpenSearch response.
+
+```mermaid
+---
+title: Nested Select Clause Class Diagram
+---
```

Comment:
On my machine it appears that the inclusion of this title section causes the graph to have a syntax error, removing these lines causes the graph to load properly.

<img width="665" alt="image" src="https://user-images.githubusercontent.com/121696727/236346922-73cb3ddd-8de1-4d6f-bb52-81f359edf5b5.png">

Reply:
If you go to the file [here](https://github.com/Bit-Quill/opensearch-project-sql/blob/7b302aff85cca44351bd2c0976c9d6582e9be525/docs/dev/sql-nested-function-select-clause.md#24-select-clause-nested-query-class-diagram), you can see that github renders the diagram correctly. I think you may need a newer version of mermaid. Using confluence which also renders this correctly I see we have version 9.4.3.